src/serdes/binary-parser.ts

```diff
@@ -1,103 +1,142 @@
-import { strict as assert } from "assert";
-import { makeClass } from "../utils/make-class";
-import { Field } from "../definitions";
-import { slice, parseBytes } from "../utils/bytes-utils";
-
-const BinaryParser = makeClass(
-  {
-    BinaryParser(buf) {
-      this._buf = parseBytes(buf, Uint8Array);
-      this._length = this._buf.length;
-      this._cursor = 0;
-    },
-    skip(n) {
-      this._cursor += n;
-    },
-    read(n, to = Uint8Array) {
-      const start = this._cursor;
-      const end = this._cursor + n;
-      assert(end <= this._buf.length);
-      this._cursor = end;
-      return slice(this._buf, start, end, to);
-    },
-    readUIntN(n) {
-      return this.read(n, Array).reduce((a, b) => (a << 8) | b) >>> 0;
-    },
-    readUInt8() {
-      return this._buf[this._cursor++];
-    },
-    readUInt16() {
-      return this.readUIntN(2);
-    },
-    readUInt32() {
-      return this.readUIntN(4);
-    },
-    pos() {
-      return this._cursor;
-    },
-    size() {
-      return this._buf.length;
-    },
-    end(customEnd) {
-      const cursor = this.pos();
-      return (
-        cursor >= this._length || (customEnd !== null && cursor >= customEnd)
+import * as assert from "assert";
+import { Field, FieldInstance } from "../definitions";
+
+class BinaryParser {
+  _buf: Buffer;
+
+  constructor(buf: string) {
```

Comment:
Please doc this noting that the input parameter is a hex string. 

Recommend renaming the parameter as `bytesHex` or `hexBytes`.